### PR TITLE
add c_order_percent for chinese market fee

### PR DESCRIPTION
**Pull request by albertandking**
Hi, Ed,
I have updated a new fee type for chinese stock market.
1. buy type should take at least 5 yuan
2. sell type should take 0.01 percent of amout stamp duty

### CODE REVIEW

File: src/pybroker/portfolio.py [hunk 1]
```diff
@@ -344,12 +344,22 @@ def __init__(
...
 
-    def _calculate_fees(self, fill_price: Decimal, shares: Decimal) -> Decimal:
+    def _calculate_fees(self, fill_price: Decimal, shares: Decimal, type: Literal["buy", "sell"]) -> Decimal:
         fees = Decimal()
         if self._fee_mode is None or self._fee_amount is None:
             return fees
         if self._fee_mode == FeeMode.ORDER_PERCENT:
             fees = self._fee_amount / Decimal(100) * fill_price * shares
+        elif self._fee_mode == FeeMode.C_ORDER_PERCENT:
+            if type == "buy":
```
Comment (edtechre) [inline src/pybroker/portfolio.py:354]: The if/else is redundant here. You can remove the repeated section to make more readable:

fees = self._fee_amount / Decimal(100) * fill_price * shares
if fees < Decimal(5.0):
     fees = Decimal(5.0)
if type == "sell":
   fees = fees + Decimal(0.0001) * fill_price * shares

File: src/pybroker/portfolio.py [hunk 2]
```diff
@@ -344,12 +344,22 @@ def __init__(
...
+                fees = self._fee_amount / Decimal(100) * fill_price * shares
+                if fees < Decimal(5.0):
+                    fees = Decimal(5.0)
+            else:
+                fees = self._fee_amount / Decimal(100) * fill_price * shares
+                if fees < Decimal(5.0):
+                    fees = Decimal(5.0)
+                fees = fees + Decimal(0.0001) * fill_price * shares
```
Comment (edtechre) [inline src/pybroker/portfolio.py:362]: Does this extra fee apply to both selling a long order and when shorting a stock?

File: src/pybroker/portfolio.py [hunk 3]
```diff
@@ -402,7 +412,7 @@ def _add_order(
         shares: Decimal,
     ) -> Order:
         self._order_id += 1
-        fees = self._calculate_fees(fill_price, shares)
+        fees = self._calculate_fees(fill_price, shares, type)
```
Comment (edtechre) [inline src/pybroker/portfolio.py:415]: This needs unit tests but I will add these myself after I merge.

File: src/pybroker/common.py [hunk 4]
```diff
@@ -145,6 +145,7 @@ class FeeMode(Enum):
     """
 
     ORDER_PERCENT = "order_percent"
+    C_ORDER_PERCENT = "c_order_percent"
```
Comment (edtechre) [inline src/pybroker/common.py:148]: This needs documentation above.

**Review by edtechre — CHANGES_REQUESTED** [review]
The only blocking is clarification on:

"Does this extra fee apply to both selling a long order and when shorting a stock?"

Otherwise I can merge and make the rest of the changes. Thanks!

**Comment by albertandking** [comment]
> The only blocking is clarification on:
> 
> 
> 
> "Does this extra fee apply to both selling a long order and when shorting a stock?"
> 
> 
> 
> Otherwise I can merge and make the rest of the changes. Thanks!

Hello Ed, in the current Chinese market, you can only go long and not short, which means you can only sell stocks that you own. You cannot sell stocks that you do not own. Therefore, there are different settings for buying and selling.

**Comment by edtechre** [comment]
Great, thank you for clarifying.

**Review by edtechre — APPROVED** [review]

**Comment by edtechre** [comment]
Hi @albertandking,

I have another question:
> fees = fees + Decimal(0.0001) * fill_price * shares

Why is Decimal(0.0001) hardcoded? Is this guaranteed to be the case across all exchanges in China?

**Comment by edtechre** [comment]
Note, a more scalable way of handling this would be to add support for setting a "FeeStructure" object that can implement custom fee logic. We could provide an implementation for the Chinese market (Shanghai stock exchange?)

**Pull request merged**